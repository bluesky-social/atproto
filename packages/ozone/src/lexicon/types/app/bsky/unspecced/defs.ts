/**
 * GENERATED CODE - DO NOT MODIFY
 */
import { type ValidationResult, BlobRef } from '@atproto/lexicon'
import { CID } from 'multiformats/cid'
import { validate as _validate } from '../../../../lexicons'
import {
  type $Typed,
  is$typed as _is$typed,
  type OmitKey,
} from '../../../../util'

const is$typed = _is$typed,
  validate = _validate
const id = 'app.bsky.unspecced.defs'

export interface SkeletonSearchPost {
  $type?: 'app.bsky.unspecced.defs#skeletonSearchPost'
  uri: string
}

const hashSkeletonSearchPost = 'skeletonSearchPost'

export function isSkeletonSearchPost<V>(v: V) {
  return is$typed(v, id, hashSkeletonSearchPost)
}

export function validateSkeletonSearchPost<V>(v: V) {
  return validate<SkeletonSearchPost & V>(v, id, hashSkeletonSearchPost)
}

export interface SkeletonSearchActor {
  $type?: 'app.bsky.unspecced.defs#skeletonSearchActor'
  did: string
}

const hashSkeletonSearchActor = 'skeletonSearchActor'

export function isSkeletonSearchActor<V>(v: V) {
  return is$typed(v, id, hashSkeletonSearchActor)
}

export function validateSkeletonSearchActor<V>(v: V) {
  return validate<SkeletonSearchActor & V>(v, id, hashSkeletonSearchActor)
}

export interface SkeletonSearchStarterPack {
  $type?: 'app.bsky.unspecced.defs#skeletonSearchStarterPack'
  uri: string
}

const hashSkeletonSearchStarterPack = 'skeletonSearchStarterPack'

export function isSkeletonSearchStarterPack<V>(v: V) {
  return is$typed(v, id, hashSkeletonSearchStarterPack)
}

export function validateSkeletonSearchStarterPack<V>(v: V) {
  return validate<SkeletonSearchStarterPack & V>(
    v,
    id,
    hashSkeletonSearchStarterPack,
  )
}

export interface TrendingTopic {
  $type?: 'app.bsky.unspecced.defs#trendingTopic'
  topic: string
  displayName?: string
  description?: string
  link: string
}

const hashTrendingTopic = 'trendingTopic'

export function isTrendingTopic<V>(v: V) {
  return is$typed(v, id, hashTrendingTopic)
}

export function validateTrendingTopic<V>(v: V) {
  return validate<TrendingTopic & V>(v, id, hashTrendingTopic)
}
