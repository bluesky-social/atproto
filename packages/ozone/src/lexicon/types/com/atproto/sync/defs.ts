/**
 * GENERATED CODE - DO NOT MODIFY
 */
import { type ValidationResult, BlobRef } from '@atproto/lexicon'
import { CID } from 'multiformats/cid'
import { validate as _validate } from '../../../../lexicons'
import {
  type $Typed,
  is$typed as _is$typed,
  type OmitKey,
} from '../../../../util'

const is$typed = _is$typed,
  validate = _validate
const id = 'com.atproto.sync.defs'

export type HostStatus =
  | 'active'
  | 'idle'
  | 'offline'
  | 'throttled'
  | 'banned'
  | (string & {})
