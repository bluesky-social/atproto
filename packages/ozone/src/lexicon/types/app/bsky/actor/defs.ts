/**
 * GENERATED CODE - DO NOT MODIFY
 */
import { ValidationResult, BlobRef } from '@atproto/lexicon'
import { CID } from 'multiformats/cid'
import { lexicons } from '../../../../lexicons'
import { $Type, is$typed } from '../../../../util'
import * as ComAtprotoLabelDefs from '../../../com/atproto/label/defs'
import * as AppBskyGraphDefs from '../graph/defs'
import * as ComAtprotoRepoStrongRef from '../../../com/atproto/repo/strongRef'

const id = 'app.bsky.actor.defs'

export interface ProfileViewBasic {
  did: string
  handle: string
  displayName?: string
  avatar?: string
  associated?: ProfileAssociated
  viewer?: ViewerState
  labels?: ComAtprotoLabelDefs.Label[]
  createdAt?: string
  [k: string]: unknown
}

export function isProfileViewBasic(v: unknown): v is ProfileViewBasic & {
  $type: $Type<'app.bsky.actor.defs', 'profileViewBasic'>
} {
  return is$typed(v, id, 'profileViewBasic')
}

export function validateProfileViewBasic(v: unknown) {
  return lexicons.validate(
    `${id}#profileViewBasic`,
    v,
  ) as ValidationResult<ProfileViewBasic>
}

export interface ProfileView {
  did: string
  handle: string
  displayName?: string
  description?: string
  avatar?: string
  associated?: ProfileAssociated
  indexedAt?: string
  createdAt?: string
  viewer?: ViewerState
  labels?: ComAtprotoLabelDefs.Label[]
  [k: string]: unknown
}

export function isProfileView(
  v: unknown,
): v is ProfileView & { $type: $Type<'app.bsky.actor.defs', 'profileView'> } {
  return is$typed(v, id, 'profileView')
}

export function validateProfileView(v: unknown) {
  return lexicons.validate(
    `${id}#profileView`,
    v,
  ) as ValidationResult<ProfileView>
}

export interface ProfileViewDetailed {
  did: string
  handle: string
  displayName?: string
  description?: string
  avatar?: string
  banner?: string
  followersCount?: number
  followsCount?: number
  postsCount?: number
  associated?: ProfileAssociated
  joinedViaStarterPack?: AppBskyGraphDefs.StarterPackViewBasic
  indexedAt?: string
  createdAt?: string
  viewer?: ViewerState
  labels?: ComAtprotoLabelDefs.Label[]
  pinnedPost?: ComAtprotoRepoStrongRef.Main
  [k: string]: unknown
}

export function isProfileViewDetailed(v: unknown): v is ProfileViewDetailed & {
  $type: $Type<'app.bsky.actor.defs', 'profileViewDetailed'>
} {
  return is$typed(v, id, 'profileViewDetailed')
}

export function validateProfileViewDetailed(v: unknown) {
  return lexicons.validate(
    `${id}#profileViewDetailed`,
    v,
  ) as ValidationResult<ProfileViewDetailed>
}

export interface ProfileAssociated {
  lists?: number
  feedgens?: number
  starterPacks?: number
  labeler?: boolean
  chat?: ProfileAssociatedChat
  [k: string]: unknown
}

export function isProfileAssociated(v: unknown): v is ProfileAssociated & {
  $type: $Type<'app.bsky.actor.defs', 'profileAssociated'>
} {
  return is$typed(v, id, 'profileAssociated')
}

export function validateProfileAssociated(v: unknown) {
  return lexicons.validate(
    `${id}#profileAssociated`,
    v,
  ) as ValidationResult<ProfileAssociated>
}

export interface ProfileAssociatedChat {
  allowIncoming: 'all' | 'none' | 'following' | (string & {})
  [k: string]: unknown
}

export function isProfileAssociatedChat(
  v: unknown,
): v is ProfileAssociatedChat & {
  $type: $Type<'app.bsky.actor.defs', 'profileAssociatedChat'>
} {
  return is$typed(v, id, 'profileAssociatedChat')
}

export function validateProfileAssociatedChat(v: unknown) {
  return lexicons.validate(
    `${id}#profileAssociatedChat`,
    v,
  ) as ValidationResult<ProfileAssociatedChat>
}

/** Metadata about the requesting account's relationship with the subject account. Only has meaningful content for authed requests. */
export interface ViewerState {
  muted?: boolean
  mutedByList?: AppBskyGraphDefs.ListViewBasic
  blockedBy?: boolean
  blocking?: string
  blockingByList?: AppBskyGraphDefs.ListViewBasic
  following?: string
  followedBy?: string
  knownFollowers?: KnownFollowers
  [k: string]: unknown
}

export function isViewerState(
  v: unknown,
): v is ViewerState & { $type: $Type<'app.bsky.actor.defs', 'viewerState'> } {
  return is$typed(v, id, 'viewerState')
}

export function validateViewerState(v: unknown) {
  return lexicons.validate(
    `${id}#viewerState`,
    v,
  ) as ValidationResult<ViewerState>
}

/** The subject's followers whom you also follow */
export interface KnownFollowers {
  count: number
  followers: ProfileViewBasic[]
  [k: string]: unknown
}

export function isKnownFollowers(v: unknown): v is KnownFollowers & {
  $type: $Type<'app.bsky.actor.defs', 'knownFollowers'>
} {
  return is$typed(v, id, 'knownFollowers')
}

export function validateKnownFollowers(v: unknown) {
  return lexicons.validate(
    `${id}#knownFollowers`,
    v,
  ) as ValidationResult<KnownFollowers>
}

export type Preferences = (
  | AdultContentPref
  | ContentLabelPref
  | SavedFeedsPref
  | SavedFeedsPrefV2
  | PersonalDetailsPref
  | FeedViewPref
  | ThreadViewPref
  | InterestsPref
  | MutedWordsPref
  | HiddenPostsPref
  | BskyAppStatePref
  | LabelersPref
  | { $type: string; [k: string]: unknown }
)[]

export interface AdultContentPref {
  enabled: boolean
  [k: string]: unknown
}

export function isAdultContentPref(v: unknown): v is AdultContentPref & {
  $type: $Type<'app.bsky.actor.defs', 'adultContentPref'>
} {
  return is$typed(v, id, 'adultContentPref')
}

export function validateAdultContentPref(v: unknown) {
  return lexicons.validate(
    `${id}#adultContentPref`,
    v,
  ) as ValidationResult<AdultContentPref>
}

export interface ContentLabelPref {
  /** Which labeler does this preference apply to? If undefined, applies globally. */
  labelerDid?: string
  label: string
  visibility: 'ignore' | 'show' | 'warn' | 'hide' | (string & {})
  [k: string]: unknown
}

export function isContentLabelPref(v: unknown): v is ContentLabelPref & {
  $type: $Type<'app.bsky.actor.defs', 'contentLabelPref'>
} {
  return is$typed(v, id, 'contentLabelPref')
}

export function validateContentLabelPref(v: unknown) {
  return lexicons.validate(
    `${id}#contentLabelPref`,
    v,
  ) as ValidationResult<ContentLabelPref>
}

export interface SavedFeed {
  id: string
  type: 'feed' | 'list' | 'timeline' | (string & {})
  value: string
  pinned: boolean
  [k: string]: unknown
}

export function isSavedFeed(
  v: unknown,
): v is SavedFeed & { $type: $Type<'app.bsky.actor.defs', 'savedFeed'> } {
  return is$typed(v, id, 'savedFeed')
}

export function validateSavedFeed(v: unknown) {
  return lexicons.validate(`${id}#savedFeed`, v) as ValidationResult<SavedFeed>
}

export interface SavedFeedsPrefV2 {
  items: SavedFeed[]
  [k: string]: unknown
}

export function isSavedFeedsPrefV2(v: unknown): v is SavedFeedsPrefV2 & {
  $type: $Type<'app.bsky.actor.defs', 'savedFeedsPrefV2'>
} {
  return is$typed(v, id, 'savedFeedsPrefV2')
}

export function validateSavedFeedsPrefV2(v: unknown) {
  return lexicons.validate(
    `${id}#savedFeedsPrefV2`,
    v,
  ) as ValidationResult<SavedFeedsPrefV2>
}

export interface SavedFeedsPref {
  pinned: string[]
  saved: string[]
  timelineIndex?: number
  [k: string]: unknown
}

export function isSavedFeedsPref(v: unknown): v is SavedFeedsPref & {
  $type: $Type<'app.bsky.actor.defs', 'savedFeedsPref'>
} {
  return is$typed(v, id, 'savedFeedsPref')
}

export function validateSavedFeedsPref(v: unknown) {
  return lexicons.validate(
    `${id}#savedFeedsPref`,
    v,
  ) as ValidationResult<SavedFeedsPref>
}

export interface PersonalDetailsPref {
  /** The birth date of account owner. */
  birthDate?: string
  [k: string]: unknown
}

export function isPersonalDetailsPref(v: unknown): v is PersonalDetailsPref & {
  $type: $Type<'app.bsky.actor.defs', 'personalDetailsPref'>
} {
  return is$typed(v, id, 'personalDetailsPref')
}

export function validatePersonalDetailsPref(v: unknown) {
  return lexicons.validate(
    `${id}#personalDetailsPref`,
    v,
  ) as ValidationResult<PersonalDetailsPref>
}

export interface FeedViewPref {
  /** The URI of the feed, or an identifier which describes the feed. */
  feed: string
  /** Hide replies in the feed. */
  hideReplies?: boolean
  /** Hide replies in the feed if they are not by followed users. */
  hideRepliesByUnfollowed: boolean
  /** Hide replies in the feed if they do not have this number of likes. */
  hideRepliesByLikeCount?: number
  /** Hide reposts in the feed. */
  hideReposts?: boolean
  /** Hide quote posts in the feed. */
  hideQuotePosts?: boolean
  [k: string]: unknown
}

export function isFeedViewPref(
  v: unknown,
): v is FeedViewPref & { $type: $Type<'app.bsky.actor.defs', 'feedViewPref'> } {
  return is$typed(v, id, 'feedViewPref')
}

export function validateFeedViewPref(v: unknown) {
  return lexicons.validate(
    `${id}#feedViewPref`,
    v,
  ) as ValidationResult<FeedViewPref>
}

export interface ThreadViewPref {
  /** Sorting mode for threads. */
  sort?:
    | 'oldest'
    | 'newest'
    | 'most-likes'
    | 'random'
    | 'hotness'
    | (string & {})
  /** Show followed users at the top of all replies. */
  prioritizeFollowedUsers?: boolean
  [k: string]: unknown
}

export function isThreadViewPref(v: unknown): v is ThreadViewPref & {
  $type: $Type<'app.bsky.actor.defs', 'threadViewPref'>
} {
  return is$typed(v, id, 'threadViewPref')
}

export function validateThreadViewPref(v: unknown) {
  return lexicons.validate(
    `${id}#threadViewPref`,
    v,
  ) as ValidationResult<ThreadViewPref>
}

export interface InterestsPref {
  /** A list of tags which describe the account owner's interests gathered during onboarding. */
  tags: string[]
  [k: string]: unknown
}

export function isInterestsPref(v: unknown): v is InterestsPref & {
  $type: $Type<'app.bsky.actor.defs', 'interestsPref'>
} {
  return is$typed(v, id, 'interestsPref')
}

export function validateInterestsPref(v: unknown) {
  return lexicons.validate(
    `${id}#interestsPref`,
    v,
  ) as ValidationResult<InterestsPref>
}

export type MutedWordTarget = 'content' | 'tag' | (string & {})

/** A word that the account owner has muted. */
export interface MutedWord {
  id?: string
  /** The muted word itself. */
  value: string
  /** The intended targets of the muted word. */
  targets: MutedWordTarget[]
  /** Groups of users to apply the muted word to. If undefined, applies to all users. */
  actorTarget: 'all' | 'exclude-following' | (string & {})
  /** The date and time at which the muted word will expire and no longer be applied. */
  expiresAt?: string
  [k: string]: unknown
}

export function isMutedWord(
  v: unknown,
): v is MutedWord & { $type: $Type<'app.bsky.actor.defs', 'mutedWord'> } {
  return is$typed(v, id, 'mutedWord')
}

export function validateMutedWord(v: unknown) {
  return lexicons.validate(`${id}#mutedWord`, v) as ValidationResult<MutedWord>
}

export interface MutedWordsPref {
  /** A list of words the account owner has muted. */
  items: MutedWord[]
  [k: string]: unknown
}

export function isMutedWordsPref(v: unknown): v is MutedWordsPref & {
  $type: $Type<'app.bsky.actor.defs', 'mutedWordsPref'>
} {
  return is$typed(v, id, 'mutedWordsPref')
}

export function validateMutedWordsPref(v: unknown) {
  return lexicons.validate(
    `${id}#mutedWordsPref`,
    v,
  ) as ValidationResult<MutedWordsPref>
}

export interface HiddenPostsPref {
  /** A list of URIs of posts the account owner has hidden. */
  items: string[]
  [k: string]: unknown
}

export function isHiddenPostsPref(v: unknown): v is HiddenPostsPref & {
  $type: $Type<'app.bsky.actor.defs', 'hiddenPostsPref'>
} {
  return is$typed(v, id, 'hiddenPostsPref')
}

export function validateHiddenPostsPref(v: unknown) {
  return lexicons.validate(
    `${id}#hiddenPostsPref`,
    v,
  ) as ValidationResult<HiddenPostsPref>
}

export interface LabelersPref {
  labelers: LabelerPrefItem[]
  [k: string]: unknown
}

export function isLabelersPref(
  v: unknown,
): v is LabelersPref & { $type: $Type<'app.bsky.actor.defs', 'labelersPref'> } {
  return is$typed(v, id, 'labelersPref')
}

export function validateLabelersPref(v: unknown) {
  return lexicons.validate(
    `${id}#labelersPref`,
    v,
  ) as ValidationResult<LabelersPref>
}

export interface LabelerPrefItem {
  did: string
  [k: string]: unknown
}

export function isLabelerPrefItem(v: unknown): v is LabelerPrefItem & {
  $type: $Type<'app.bsky.actor.defs', 'labelerPrefItem'>
} {
  return is$typed(v, id, 'labelerPrefItem')
}

export function validateLabelerPrefItem(v: unknown) {
  return lexicons.validate(
    `${id}#labelerPrefItem`,
    v,
  ) as ValidationResult<LabelerPrefItem>
}

/** A grab bag of state that's specific to the bsky.app program. Third-party apps shouldn't use this. */
export interface BskyAppStatePref {
  activeProgressGuide?: BskyAppProgressGuide
  /** An array of tokens which identify nudges (modals, popups, tours, highlight dots) that should be shown to the user. */
  queuedNudges?: string[]
  /** Storage for NUXs the user has encountered. */
  nuxs?: Nux[]
  [k: string]: unknown
}

export function isBskyAppStatePref(v: unknown): v is BskyAppStatePref & {
  $type: $Type<'app.bsky.actor.defs', 'bskyAppStatePref'>
} {
  return is$typed(v, id, 'bskyAppStatePref')
}

export function validateBskyAppStatePref(v: unknown) {
  return lexicons.validate(
    `${id}#bskyAppStatePref`,
    v,
  ) as ValidationResult<BskyAppStatePref>
}

/** If set, an active progress guide. Once completed, can be set to undefined. Should have unspecced fields tracking progress. */
export interface BskyAppProgressGuide {
  guide: string
  [k: string]: unknown
}

export function isBskyAppProgressGuide(
  v: unknown,
): v is BskyAppProgressGuide & {
  $type: $Type<'app.bsky.actor.defs', 'bskyAppProgressGuide'>
} {
  return is$typed(v, id, 'bskyAppProgressGuide')
}

export function validateBskyAppProgressGuide(v: unknown) {
  return lexicons.validate(
    `${id}#bskyAppProgressGuide`,
    v,
  ) as ValidationResult<BskyAppProgressGuide>
}

/** A new user experiences (NUX) storage object */
export interface Nux {
  id: string
  completed: boolean
  /** Arbitrary data for the NUX. The structure is defined by the NUX itself. Limited to 300 characters. */
  data?: string
  /** The date and time at which the NUX will expire and should be considered completed. */
  expiresAt?: string
  [k: string]: unknown
}

export function isNux(
  v: unknown,
): v is Nux & { $type: $Type<'app.bsky.actor.defs', 'nux'> } {
  return is$typed(v, id, 'nux')
}

export function validateNux(v: unknown) {
  return lexicons.validate(`${id}#nux`, v) as ValidationResult<Nux>
}
