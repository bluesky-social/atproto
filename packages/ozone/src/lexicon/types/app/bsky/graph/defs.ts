/**
 * GENERATED CODE - DO NOT MODIFY
 */
import { ValidationResult, BlobRef } from '@atproto/lexicon'
import { CID } from 'multiformats/cid'
import { lexicons } from '../../../../lexicons'
import { $Type, is$typed } from '../../../../util'
import * as ComAtprotoLabelDefs from '../../../com/atproto/label/defs'
import * as AppBskyActorDefs from '../actor/defs'
import * as AppBskyRichtextFacet from '../richtext/facet'
import * as AppBskyFeedDefs from '../feed/defs'

const id = 'app.bsky.graph.defs'

export interface ListViewBasic {
  uri: string
  cid: string
  name: string
  purpose: ListPurpose
  avatar?: string
  listItemCount?: number
  labels?: ComAtprotoLabelDefs.Label[]
  viewer?: ListViewerState
  indexedAt?: string
  [k: string]: unknown
}

export function isListViewBasic(v: unknown): v is ListViewBasic & {
  $type: $Type<'app.bsky.graph.defs', 'listViewBasic'>
} {
  return is$typed(v, id, 'listViewBasic')
}

export function validateListViewBasic(v: unknown) {
  return lexicons.validate(
    `${id}#listViewBasic`,
    v,
  ) as ValidationResult<ListViewBasic>
}

export interface ListView {
  uri: string
  cid: string
  creator: AppBskyActorDefs.ProfileView
  name: string
  purpose: ListPurpose
  description?: string
  descriptionFacets?: AppBskyRichtextFacet.Main[]
  avatar?: string
  listItemCount?: number
  labels?: ComAtprotoLabelDefs.Label[]
  viewer?: ListViewerState
  indexedAt: string
  [k: string]: unknown
}

export function isListView(
  v: unknown,
): v is ListView & { $type: $Type<'app.bsky.graph.defs', 'listView'> } {
  return is$typed(v, id, 'listView')
}

export function validateListView(v: unknown) {
  return lexicons.validate(`${id}#listView`, v) as ValidationResult<ListView>
}

export interface ListItemView {
  uri: string
  subject: AppBskyActorDefs.ProfileView
  [k: string]: unknown
}

export function isListItemView(
  v: unknown,
): v is ListItemView & { $type: $Type<'app.bsky.graph.defs', 'listItemView'> } {
  return is$typed(v, id, 'listItemView')
}

export function validateListItemView(v: unknown) {
  return lexicons.validate(
    `${id}#listItemView`,
    v,
  ) as ValidationResult<ListItemView>
}

export interface StarterPackView {
  uri: string
  cid: string
  record: {}
  creator: AppBskyActorDefs.ProfileViewBasic
  list?: ListViewBasic
  listItemsSample?: ListItemView[]
  feeds?: AppBskyFeedDefs.GeneratorView[]
  joinedWeekCount?: number
  joinedAllTimeCount?: number
  labels?: ComAtprotoLabelDefs.Label[]
  indexedAt: string
  [k: string]: unknown
}

export function isStarterPackView(v: unknown): v is StarterPackView & {
  $type: $Type<'app.bsky.graph.defs', 'starterPackView'>
} {
  return is$typed(v, id, 'starterPackView')
}

export function validateStarterPackView(v: unknown) {
  return lexicons.validate(
    `${id}#starterPackView`,
    v,
  ) as ValidationResult<StarterPackView>
}

export interface StarterPackViewBasic {
  uri: string
  cid: string
  record: {}
  creator: AppBskyActorDefs.ProfileViewBasic
  listItemCount?: number
  joinedWeekCount?: number
  joinedAllTimeCount?: number
  labels?: ComAtprotoLabelDefs.Label[]
  indexedAt: string
  [k: string]: unknown
}

export function isStarterPackViewBasic(
  v: unknown,
): v is StarterPackViewBasic & {
  $type: $Type<'app.bsky.graph.defs', 'starterPackViewBasic'>
} {
  return is$typed(v, id, 'starterPackViewBasic')
}

export function validateStarterPackViewBasic(v: unknown) {
  return lexicons.validate(
    `${id}#starterPackViewBasic`,
    v,
  ) as ValidationResult<StarterPackViewBasic>
}

export type ListPurpose =
  | 'app.bsky.graph.defs#modlist'
  | 'app.bsky.graph.defs#curatelist'
  | 'app.bsky.graph.defs#referencelist'
  | (string & {})

/** A list of actors to apply an aggregate moderation action (mute/block) on. */
export const MODLIST = `${id}#modlist`
/** A list of actors used for curation purposes such as list feeds or interaction gating. */
export const CURATELIST = `${id}#curatelist`
/** A list of actors used for only for reference purposes such as within a starter pack. */
export const REFERENCELIST = `${id}#referencelist`

export interface ListViewerState {
  muted?: boolean
  blocked?: string
  [k: string]: unknown
}

export function isListViewerState(v: unknown): v is ListViewerState & {
  $type: $Type<'app.bsky.graph.defs', 'listViewerState'>
} {
  return is$typed(v, id, 'listViewerState')
}

export function validateListViewerState(v: unknown) {
  return lexicons.validate(
    `${id}#listViewerState`,
    v,
  ) as ValidationResult<ListViewerState>
}

/** indicates that a handle or DID could not be resolved */
export interface NotFoundActor {
  actor: string
  notFound: true
  [k: string]: unknown
}

export function isNotFoundActor(v: unknown): v is NotFoundActor & {
  $type: $Type<'app.bsky.graph.defs', 'notFoundActor'>
} {
  return is$typed(v, id, 'notFoundActor')
}

export function validateNotFoundActor(v: unknown) {
  return lexicons.validate(
    `${id}#notFoundActor`,
    v,
  ) as ValidationResult<NotFoundActor>
}

/** lists the bi-directional graph relationships between one actor (not indicated in the object), and the target actors (the DID included in the object) */
export interface Relationship {
  did: string
  /** if the actor follows this DID, this is the AT-URI of the follow record */
  following?: string
  /** if the actor is followed by this DID, contains the AT-URI of the follow record */
  followedBy?: string
  [k: string]: unknown
}

export function isRelationship(
  v: unknown,
): v is Relationship & { $type: $Type<'app.bsky.graph.defs', 'relationship'> } {
  return is$typed(v, id, 'relationship')
}

export function validateRelationship(v: unknown) {
  return lexicons.validate(
    `${id}#relationship`,
    v,
  ) as ValidationResult<Relationship>
}
