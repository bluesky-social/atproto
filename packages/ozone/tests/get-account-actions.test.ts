import AtpAgent, {
  ToolsOzoneHistoryGetAccountActions,
  ToolsOzoneHistoryGetSubjectHistory,
} from '@atproto/api'
import {
  ModeratorClient,
  SeedClient,
  TestNetwork,
  basicSeed,
} from '@atproto/dev-env'
import {
  REASONMISLEADING,
  REASONSPAM,
} from '../src/lexicon/types/com/atproto/moderation/defs'
import { MODACTIONLABEL } from '../src/lexicon/types/tools/ozone/history/defs'
import { forSnapshot } from './_util'

type UserWithAgent = {
  agent?: AtpAgent
  token: string
}

describe('get-account-actions', () => {
  let network: TestNetwork
  let sc: SeedClient
  const alice: UserWithAgent = {
    token: '',
  }
  const carol: UserWithAgent = {
    token: '',
  }
  const bob: UserWithAgent = {
    token: '',
  }
  let modClient: ModeratorClient

  const seedReports = async () => {
    await sc.createReport({
      subject: {
        $type: 'com.atproto.admin.defs#repoRef',
        did: sc.dids.carol,
      },
      reasonType: REASONMISLEADING,
      reason: 'alice says carol is bad',
      reportedBy: sc.dids.alice,
    })
    await sc.createReport({
      subject: {
        $type: 'com.atproto.repo.strongRef',
        uri: sc.posts[sc.dids.carol][0].ref.uriStr,
        cid: sc.posts[sc.dids.carol][0].ref.cidStr,
      },
      reasonType: REASONMISLEADING,
      reason: 'alice says carol cannot post',
      reportedBy: sc.dids.alice,
    })
    await sc.createReport({
      subject: {
        $type: 'com.atproto.admin.defs#repoRef',
        did: sc.dids.bob,
      },
      reasonType: REASONSPAM,
      reason: 'alice says bob is bad',
      reportedBy: sc.dids.alice,
    })
    await sc.createReport({
      subject: {
        $type: 'com.atproto.repo.strongRef',
        uri: sc.posts[sc.dids.bob][0].ref.uriStr,
        cid: sc.posts[sc.dids.bob][0].ref.cidStr,
      },
      reasonType: REASONSPAM,
      reason: 'alice says bob cannot post',
      reportedBy: sc.dids.alice,
    })
    await sc.createReport({
      subject: {
        $type: 'com.atproto.repo.strongRef',
        uri: sc.posts[sc.dids.bob][0].ref.uriStr,
        cid: sc.posts[sc.dids.bob][0].ref.cidStr,
      },
      reasonType: REASONSPAM,
      reason: 'carol also says bob cannot post',
      reportedBy: sc.dids.carol,
    })
  }

  const getActions = async (
    user: UserWithAgent,
    params: ToolsOzoneHistoryGetAccountActions.QueryParams,
  ) => {
    if (!user.agent) {
      throw new Error('User agent not set')
    }
    const { data } = await user.agent.tools.ozone.history.getAccountActions(
      params,
      {
        headers: {
          Authorization: `Bearer ${user.token}`,
          'atproto-proxy': `${network.ozone.ctx.cfg.service.did}#atproto_labeler`,
        },
      },
    )

    return data.subjects
  }

  const getSubjectHistory = async (
    user: UserWithAgent,
    params: ToolsOzoneHistoryGetSubjectHistory.QueryParams,
  ) => {
    if (!user.agent) {
      throw new Error('User agent not set')
    }
    const { data } = await user.agent.tools.ozone.history.getSubjectHistory(
      params,
      {
        headers: {
          Authorization: `Bearer ${user.token}`,
          'atproto-proxy': `${network.ozone.ctx.cfg.service.did}#atproto_labeler`,
        },
      },
    )

    return data.events
  }

  beforeAll(async () => {
    network = await TestNetwork.create({
      dbPostgresSchema: 'ozone_get_account_actions',
    })
    sc = network.getSeedClient()
    modClient = network.ozone.getModClient()
    await basicSeed(sc)
    await network.processAll()
    await seedReports()
    alice.agent = network.pds.getClient()
    const { data: alicesLogin } = await alice.agent.login({
      identifier: sc.accounts[sc.dids.alice].handle,
      password: sc.accounts[sc.dids.alice].password,
    })
    carol.agent = network.pds.getClient()
    const { data: carolsLogin } = await carol.agent.login({
      identifier: sc.accounts[sc.dids.carol].handle,
      password: sc.accounts[sc.dids.carol].password,
    })
    bob.agent = network.pds.getClient()
    const { data: bobsLogin } = await bob.agent.login({
      identifier: sc.accounts[sc.dids.bob].handle,
      password: sc.accounts[sc.dids.bob].password,
    })
    alice.token = alicesLogin.accessJwt
    carol.token = carolsLogin.accessJwt
    carol.token = carolsLogin.accessJwt
    bob.token = bobsLogin.accessJwt
  })

  afterAll(async () => {
    await network.close()
  })

  it("returns all subjects that were actioned on the caller's account", async () => {
    const [actionsOnCarol, actionsOnBob] = await Promise.all([
      getActions(carol, {}),
      getActions(bob, {}),
    ])

    expect(forSnapshot(actionsOnCarol)).toMatchSnapshot()
    expect(forSnapshot(actionsOnBob)).toMatchSnapshot()
  })

  it('returns updated subject status after mod action', async () => {
    const bobsPostSubject = {
      $type: 'com.atproto.repo.strongRef',
      uri: sc.posts[sc.dids.bob][0].ref.uriStr,
      cid: sc.posts[sc.dids.bob][0].ref.cidStr,
    }
    await modClient.emitEvent(
      {
        event: {
          $type: 'tools.ozone.moderation.defs#modEventLabel',
          createLabelVals: ['spam'],
          negateLabelVals: [],
        },
        subject: bobsPostSubject,
      },
      'admin',
    )
    await modClient.emitEvent(
      {
        event: {
          $type: 'tools.ozone.moderation.defs#modEventAcknowledge',
        },
        subject: bobsPostSubject,
      },
      'admin',
    )

    const actionsOnBob = await getActions(bob, {})

    const actionOnBobsPost = actionsOnBob.find(
      (item) => item.subject === sc.posts[sc.dids.bob][0].ref.uriStr,
    )?.modAction

    expect(actionOnBobsPost).toEqual(MODACTIONLABEL)
  })

  it('shows subject history for a specific subject', async () => {
    const history = await getSubjectHistory(bob, {
      subject: sc.posts[sc.dids.bob][0].ref.uriStr,
    })
    expect(forSnapshot(history)).toMatchSnapshot()
  })
  it('does not show subject history of a different user', async () => {
    await expect(
      getSubjectHistory(carol, {
        subject: sc.posts[sc.dids.bob][0].ref.uriStr,
      }),
    ).rejects.toThrow('Unauthorized')
  })
})
