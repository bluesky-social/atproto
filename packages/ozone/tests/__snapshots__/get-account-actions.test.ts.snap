// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`get-account-actions returns all subjects that were actioned on the caller's account 1`] = `
Array [
  Object {
    "createdAt": "1970-01-01T00:00:00.000Z",
    "modAction": "tools.ozone.history.defs#modActionPending",
    "status": "active",
    "subject": "record(0)",
  },
  Object {
    "createdAt": "1970-01-01T00:00:00.000Z",
    "modAction": "tools.ozone.history.defs#modActionPending",
    "status": "active",
    "subject": "user(0)",
  },
]
`;

exports[`get-account-actions returns all subjects that were actioned on the caller's account 2`] = `
Array [
  Object {
    "createdAt": "1970-01-01T00:00:00.000Z",
    "modAction": "tools.ozone.history.defs#modActionPending",
    "status": "active",
    "subject": "record(0)",
  },
  Object {
    "createdAt": "1970-01-01T00:00:00.000Z",
    "modAction": "tools.ozone.history.defs#modActionPending",
    "status": "active",
    "subject": "user(0)",
  },
]
`;

exports[`get-account-actions shows subject history for a specific subject 1`] = `
Array [
  Object {
    "createdAt": "1970-01-01T00:00:00.000Z",
    "event": Object {
      "$type": "tools.ozone.history.defs#eventResolve",
    },
    "isAutomated": false,
    "subject": "record(0)",
  },
  Object {
    "createdAt": "1970-01-01T00:00:00.000Z",
    "event": Object {
      "$type": "tools.ozone.history.defs#eventLabel",
      "createLabelVals": Array [
        "spam",
      ],
      "negateLabelVals": Array [],
    },
    "isAutomated": false,
    "subject": "record(0)",
  },
  Object {
    "createdAt": "1970-01-01T00:00:00.000Z",
    "event": Object {
      "$type": "tools.ozone.history.defs#eventReport",
      "reportType": "com.atproto.moderation.defs#reasonSpam",
    },
    "isAutomated": false,
    "subject": "record(0)",
  },
  Object {
    "createdAt": "1970-01-01T00:00:00.000Z",
    "event": Object {
      "$type": "tools.ozone.history.defs#eventReport",
      "reportType": "com.atproto.moderation.defs#reasonSpam",
    },
    "isAutomated": false,
    "subject": "record(0)",
  },
]
`;
